### ISSUE TITLE
.NET 6: `_ResolveAssemblies` ignores user's `AppendRuntimeIdentifierToOutputPath`

### ISSUE THREAD

**Issue by bddckr** [xamarin/xamarin-android#6158]
In my projects/solution I'm overriding the various `*OutputPath` MSBuild properties to move all output into a root `artifacts` directory.
```xml
<OutputPath>$(ArtifactsBuildPath)bin/$(Configuration)/$(TargetFramework)/$(MSBuildProjectName)/</OutputPath>
<OutDir>$(OutputPath)</OutDir>
<BaseIntermediateOutputPath>$(ArtifactsBuildPath)obj/base/$(TargetFramework)/$(MSBuildProjectName)/</BaseIntermediateOutputPath>
<IntermediateOutputPath>$(ArtifactsBuildPath)obj/$(Configuration)/$(TargetFramework)/$(MSBuildProjectName)/</IntermediateOutputPath>
<IntDir>$(IntermediateOutputPath)</IntDir>
<AppendTargetFrameworkToOutputPath>false</AppendTargetFrameworkToOutputPath>
<AppendRuntimeIdentifierToOutputPath>false</AppendRuntimeIdentifierToOutputPath>
```

This has been working without issues until I started testing .NET 6. On preview 6, when creating an SDK-style project as a bindings library project for an `.aar`, the following error is encountered with such a customized setup:
```
"C:\Projects\App\App.csproj" (Build target) (1:7) ->
"C:\Projects\App\App.csproj" (_ComputeFilesToPublishForRuntimeIdentifiers target) (1:8) ->
"C:\Projects\Bindings\Bindings.csproj" (default target) (4:10) ->
(CopyFilesToOutputDirectory target) ->
 C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(4501,5): error MSB3030: Could not copy the file "C:\Projects\artifacts\build\obj\Debug\net6.0-android\android.30-arm64\Bindings\android.30-arm64\Bindings.dll" because it was not found. [C:\Projects\Bindings\Bindings.csproj]
```
As you can see the path is mentioning the runtime identifier two times, once more after the project name. This is not part of my customized setup and not desired. On the other projects in my solution (SDK-style and legacy projects for XA apps and libraries) no issue is encountered.

---

Turns out this is due to the following ignoring the property and instead always setting it to `true` here:
https://github.com/xamarin/xamarin-android/blob/c929289a7b77ecbc90c3e10aa55f9f3027ed3345/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.AssemblyResolution.targets#L65

Deleting this line resolves the issue in the affected project, while also keeping all other projects working without issues. By deleting it, the `_ResolveAssemblies` target uses the value of `AppendRuntimeIdentifierToOutputPath` that I've set - `false`.

**Comment by jonathanpeppers** [comment]
The problem with forcing `AppendRuntimeIdentifierToOutputPath=false`. The linker runs for each `$(RuntimeIdentifier)` and so this could overwrite files. By default an app builds for `android-arm64` and `android-x86`. And so you're probably ending up with files for the wrong architecture in your app? Does the app still "work" on emulators and devices? I suspect it will crash on one architecture.

Why do you need to override `$(IntermediateOutputPath)` exactly? It seems like you would be better off to only override `$(OutputPath)` -- choosing where your build output is produced makes sense.

**Comment by bddckr** [comment]
> Why do you need to override $(IntermediateOutputPath) exactly?

It just simplifies it when I can tell my team members to wipe a specific folder whenever they encounter a build issue that can be resolved by wiping all `obj` and `bin` folders.

> The problem with forcing AppendRuntimeIdentifierToOutputPath=false [...]

Oh sorry - that's not what my intention is!

Since this is using the `MSBuild` task all specified `Properties` apply as _global_ properties to the project that the task is executing a target on. Global properties cannot be changed, which means my MSBuild configuration to set `AppendRuntimeIdentifierToOutputPath` to `false` does not work. (I can see this in the binlog - it keeps saying that the global property cannot be changed according to MSBuild Log Viewer.)

Initially I've just changed the mentioned line to `AppendRuntimeIdentifierToOutputPath=$(AppendRuntimeIdentifierToOutputPath)` which works. Then I noticed that the MSBuild task probably does pass through existing properties as is as its default behavior, so I also tried with removing the line altogether. Both ways work as far as I can see.

So to recap: My suggestion is to _remove_ the line to allow the default to be passed through. If the default has been changed by the user, it is now used. If the user has forced the setting to `false` and has more than one runtime-identifier in use, that issue is on them, right? This is an issue in other SDK-style projects (besides the net .NET 6 Android project) and has been for a while as far as I know.

**Comment by jonathanpeppers** [comment]
@bddckr can you make a small sample that reproduces this issue? Thanks.

**Comment by bddckr** [comment]
[AppendRuntimeIdentifierToOutputPathIssue.zip](https://github.com/xamarin/xamarin-android/files/6920695/AppendRuntimeIdentifierToOutputPathIssue.zip)

Reproduction steps:
1. Run
    ```powershell
    Get-ChildItem .\ -include bin,obj -Recurse | ForEach-Object ($_) { Remove-Item $_.FullName -Force -Recurse }
    msbuild -nologo -maxcpucount -restore -target:Build App -p:Configuration=Debug
  	```
  	Note:
  	* For the first command it should be enough to just delete the `artifacts` folder, as that's what these projects are configured to use. However, just to ensure this test does not use any leftovers of some build output from runs with changed settings, this command just deletes all output folders, including the ones in `artifacts`.
  	* You can run `dotnet build` instead of `msbuild ...` here - I'm just using MSBuild.exe because I have legacy projects in my regular solution that require it.
2. Observe that it fails with
    ```
       "C:\Projects\AppendRuntimeIdentifierToOutputPathIssue\App\App.csproj" (Build target) (1:7) ->
       "C:\Projects\AppendRuntimeIdentifierToOutputPathIssue\App\App.csproj" (_ComputeFilesToPublishForRuntimeIdentifiers target) (1:8) ->
       "C:\Projects\AppendRuntimeIdentifierToOutputPathIssue\Bindings\Bindings.csproj" (default target) (2:10) ->
       (CopyFilesToOutputDirectory target) ->
         C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets(4501,5): error MSB3030: Could not copy the file "C:\Projects\AppendRuntimeIdentifierToOutputPathIssue\/a
       rtifacts/obj/Debug/net6.0-android/android.30-arm64/Bindings/android.30-arm64\Bindings.dll" because it was not found. [C:\Projects\AppendRuntimeIdentifierToOutputPathIssue\Bindings\Bindings.csproj]
    ```
3. Remove this line in your local copy of it (at `C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\30.0.100-preview.6.62\targets\Microsoft.Android.Sdk.AssemblyResolution.targets`).
    https://github.com/xamarin/xamarin-android/blob/c929289a7b77ecbc90c3e10aa55f9f3027ed3345/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.AssemblyResolution.targets#L65
4. Run the above commands again.
5. Observe that the build succeeds.
6. Repeat steps 3 to 5 again, this time *changing* the line to say `AppendRuntimeIdentifierToOutputPath=$(AppendRuntimeIdentifierToOutputPath);` instead. The line should be OK to be removed as long as the default behavior of the `MSBuild` task is to pass on the currently defined properties.

Note: It looks like the whole solution builds (and rebuilds) without any issues in Visual Studio.

**Comment by MagicAndre1981** [comment]
> Why do you need to override `$(IntermediateOutputPath)` exactly? 

I haven't seen the issue mentioned here, but I always do this to to move my obj folder into root to able to delete obj folders easier instead of going into each project and find all obj folders. I also move the nuget restore data into an extra folder to only delete them instead of whole obj folder.

**Comment by bddckr** [comment]
For anyone following along: This is still an issue with .NET 6 RC 1.

To manually work around the issue you have to edit `C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\31.0.100-rc.1.12\targets\Microsoft.Android.Sdk.AssemblyResolution.targets` after you installed the workload (e.g. `dotnet workload install android`, but you could also install it via the MAUI workloads): Remove `AppendRuntimeIdentifierToOutputPath=true` in line 75.

@jonathanpeppers Any chance this can make it into RC 2 so dev machines and CI machines alike don't need this manual workaround?

**Comment by jonathanpeppers** [comment]
@bddckr if you have a fix in mind, a pull request is welcome. I would add a test [here](https://github.com/xamarin/xamarin-android/blob/main/src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/XASdkTests.cs) using a custom `$(IntermediateOutputPath)` and other settings.

Other issues are just higher priority right now, and we have until [Q2 2022](https://devblogs.microsoft.com/dotnet/update-on-dotnet-maui/).

**Comment by bddckr** [comment]
I'll look into that as soon as I can. Thanks for the pointer as to where to start!

---

TIL that MAUI is not just the UI framework, but also means the whole effort of getting Xamarin "moved over" into the new world order in .NET. I've never understood it that way and was instead under the impression that the "Xamarin Native" stuff is independent from MAUI. Especially since it spells out as "Multi-platform App UI" - UI, not "mobile runtimes, native API bindings etc.".

That being said, thanks for pointing that out and best of luck to you and the team to continue the great work. Now hopefully on a little bit stressful timeline. 🤞 I'll update this issue once I can.

**Comment by sahin52** [comment]
Can you prioritize this?
